txnsync/bitmask.go

```diff
@@ -0,0 +1,215 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"bytes"
+	"errors"
+)
+
+var errIndexNotFound = errors.New("invalid bitmask: index not found")
+var errInvalidBitmaskType = errors.New("invalid bitmask type")
+
+//msgp:allocbound bitmask maxBitmaskSize
+type bitmask []byte
+
+// assumed to be in mode 0, sets bit at index to 1
+func (b *bitmask) setBit(index int) {
+	byteIndex := index/8 + 1
+	(*b)[byteIndex] |= 1 << (index % 8)
+}
+
+// trimBitmask compresses the bitmask into one of the 4 types:
+// type 0: input bitmask bit pos x b -> output bitmask bit pos x b
+// type 1: input bitmask bit pos x b -> output bitmask bit pos x !b
+// type 2: stores the positions of bits where the bit value is 1
+//         input bitmask first bit 1 at pos A, second bit 1 at pos B, ...
+//         output bitmask stores A, B-A, ...
+// type 3: same as type 2, but stores the positons where the bit is 0
+func (b *bitmask) trimBitmask(entries int) {
+	if *b == nil {
+		return
+	}
+	lastExists := 0
+	lastNotExists := 0
+	numExists := 0
+	for i := 0; i < entries; i++ {
+		byteIndex := i/8 + 1
+		if (*b)[byteIndex]&(1<<(i%8)) != 0 {
+			lastExists = i
+			numExists++
+		} else {
+			lastNotExists = i
+		}
+	}
+	bitmaskType := 0
+	bestSize := bytesNeededBitmask(lastExists)
+	if bestSize > bytesNeededBitmask(lastNotExists) {
+		bitmaskType = 1
+		bestSize = bytesNeededBitmask(lastNotExists)
+	}
+	if bestSize > numExists*2+1 {
+		bitmaskType = 2
+		bestSize = numExists*2 + 1
+	}
+	if bestSize > (entries-numExists)*2+1 {
+		bitmaskType = 3
+		bestSize = (entries-numExists)*2 + 1
+	}
+	switch bitmaskType {
+	case 1:
+		(*b)[0] = 1
+		for i := range *b {
+			if i != 0 {
+				(*b)[i] = 255 - (*b)[i] // invert bits
+			}
+		}
+	case 2:
+		newBitmask := make(bitmask, 1, bestSize)
+		newBitmask[0] = 2
+		last := 0
+		for i := 0; i < entries; i++ {
+			byteIndex := i/8 + 1
+			if (*b)[byteIndex]&(1<<(i%8)) != 0 {
+				diff := i - last
+				newBitmask = append(newBitmask, byte(diff/256), byte(diff%256))
+				last = i
+			}
+		}
+		*b = newBitmask
+		return
+	case 3:
+		newBitmask := make(bitmask, 1, bestSize)
+		newBitmask[0] = 3
+		last := 0
+		for i := 0; i < entries; i++ {
+			byteIndex := i/8 + 1
+			if (*b)[byteIndex]&(1<<(i%8)) == 0 {
+				diff := i - last
+				newBitmask = append(newBitmask, byte(diff/256), byte(diff%256))
+				last = i
+			}
+		}
+		*b = newBitmask
+		return
+	default:
+	}
+
+	*b = bytes.TrimRight(*b, "\x00")
+}
+
+// iterate through the elements of bitmask without expanding it.
+// If more than maxIndex set bits are found, return error. For each set value, call callback.
+func (b *bitmask) iterate(entries int, maxIndex int, callback func(int, int) error) error {
```

Comment:
I'm not sure that I'm following you. how does the effective max index ( which depends on the encoding of the bitmask ) related to the nextSlice ( which gets the next element in a slice ) ?

Reply:
What you're saying here is correct for that above use case, but would be safe on others:
```golang
	err = stub.BitmaskKeys.iterate(int(stub.TotalTransactionsCount), len(stub.VoteKeyDilution), func(i int, index int) error {
		signedTxns[i].Txn.VoteKeyDilution = stub.VoteKeyDilution[index]
		err := nextSlice(&stub.VotePK, signedTxns[i].Txn.VotePK[:], crypto.DigestSize)
		if err != nil {
			return err
		}
		return nextSlice(&stub.SelectionPK, signedTxns[i].Txn.SelectionPK[:], crypto.DigestSize)
	})
```

Reply:
nevermind that; I think you're correct here. The above isn't an issue since we have 
```golang
	// should all have same number of elements
	if len(stub.VotePK)/crypto.DigestSize != len(stub.VoteKeyDilution) || len(stub.SelectionPK)/crypto.DigestSize != len(stub.VoteKeyDilution) {
		return errDataMissing
	}
```